### ISSUE TITLE
"ember build --prod" fails occasionally with rmdir, unlink or open error

### ISSUE THREAD

**Issue by rajcoss** [ember-cli/ember-cli#5827]
Here are the steps to reproduce this...

```
mkdir -p ~/dev
cd ~/dev

ember new test_app
cd test_app
mkdir public/myfolder
echo "hello" > public/myfolder/abc.txt
ember build
ember build --prod
```

Sometimes, you might not get an error on the last line `ember build --prod`. In that case, run `ember build && ember build --prod` a couple of times. The problem shows up about 50% of the times.

The errors can be different... here are some of them
```
$ ember build && ember build --prod
Could not start watchman; falling back to NodeWatcher for file system events.
Visit http://ember-cli.com/user-guide/#watchman for more info.
Build failed.
EISDIR: illegal operation on a directory, unlink '/home/rajcoss/dev/test_app/dist/assets'
Error: EISDIR: illegal operation on a directory, unlink '/home/rajcoss/dev/test_app/dist/assets'
    at Error (native)
    at Object.fs.unlinkSync (fs.js:980:18)
    at TreeSync.<anonymous> (/home/rajcoss/dev/test_app/node_modules/ember-cli/node_modules/tree-sync/index.js:71:21)
    at Array.forEach (native)
    at TreeSync.sync (/home/rajcoss/dev/test_app/node_modules/ember-cli/node_modules/tree-sync/index.js:50:14)
    at Class.module.exports.Task.extend.copyToOutputPath (/home/rajcoss/dev/test_app/node_modules/ember-cli/lib/models/builder.js:124:10)
    at /home/rajcoss/dev/test_app/node_modules/ember-cli/lib/models/builder.js:132:21
    at lib$rsvp$$internal$$tryCatch (/home/rajcoss/dev/test_app/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:1036:16)
    at lib$rsvp$$internal$$invokeCallback (/home/rajcoss/dev/test_app/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:1048:17)
    at /home/rajcoss/dev/test_app/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:331:11


$ ember build && ember build --prod
Could not start watchman; falling back to NodeWatcher for file system events.
Visit http://ember-cli.com/user-guide/#watchman for more info.
Build failed.
ENOENT: no such file or directory, open '/home/rajcoss/dev/test_app/dist/assets/failed.png'
Error: ENOENT: no such file or directory, open '/home/rajcoss/dev/test_app/dist/assets/failed.png'
    at Error (native)
    at Object.fs.openSync (fs.js:584:18)
    at Object.fs.writeFileSync (fs.js:1224:33)
    at TreeSync.<anonymous> (/home/rajcoss/dev/test_app/node_modules/ember-cli/node_modules/tree-sync/index.js:59:19)
    at Array.forEach (native)
    at TreeSync.sync (/home/rajcoss/dev/test_app/node_modules/ember-cli/node_modules/tree-sync/index.js:50:14)
    at Class.module.exports.Task.extend.copyToOutputPath (/home/rajcoss/dev/test_app/node_modules/ember-cli/lib/models/builder.js:124:10)
    at /home/rajcoss/dev/test_app/node_modules/ember-cli/lib/models/builder.js:132:21
    at lib$rsvp$$internal$$tryCatch (/home/rajcoss/dev/test_app/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:1036:16)
    at lib$rsvp$$internal$$invokeCallback (/home/rajcoss/dev/test_app/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:1048:17)
```

Output from ember version --verbose:
```
Could not start watchman; falling back to NodeWatcher for file system events.
Visit http://ember-cli.com/user-guide/#watchman for more info.
ember-cli: 2.5.0
http_parser: 2.6.0
node: 5.3.0
v8: 4.6.85.31
uv: 1.8.0
zlib: 1.2.8
ares: 1.10.1-DEV
icu: 56.1
modules: 47
openssl: 1.0.2e
os: linux x64
```

We reproduced this on ubuntu and mac independently.

We got around this error by `rm -rf dist` before running `ember build --prod`.

**Comment by stefanpenner** [comment]
I'm pretty sure the issue is the same as the one we see in rebuilds sometimes. @hjdivad most likely an issue with how we handle directories in fs-tree-diff.

**Comment by nathanhammond** [comment]
With a test of 20 iterations on the most recent version I am no longer able to reproduce. I'm closing this as fixed.

```sh
for run in {1..10}
do
  ember build && ember build --prod
done
```

**Comment by hjdivad** [comment]
if it's actually the same issue @stefanpenner  mentioned then it would have been fixed in `fs-tree-diff@0.5.0`
